tests/test_extents.cpp

```diff
@@ -0,0 +1,176 @@
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <experimental/mdspan>
+
+#include <gtest/gtest.h>
+
+namespace stdex = std::experimental;
+
+template <class> struct TestExtents;
+template <ptrdiff_t... Extents, ptrdiff_t... DynamicSizes>
+struct TestExtents<std::tuple<
+  stdex::extents<Extents...>,
+  std::integer_sequence<ptrdiff_t, DynamicSizes...>
+>> : public ::testing::Test {
+  using extents_type = stdex::extents<Extents...>;
+  const std::array<ptrdiff_t, sizeof...(Extents)> static_sizes { Extents... };
+  const std::array<ptrdiff_t, sizeof...(DynamicSizes)> dyn_sizes { DynamicSizes... };
+  extents_type exts { DynamicSizes... };
+};
+
+template <ptrdiff_t... Ds>
+using _sizes = std::integer_sequence<ptrdiff_t, Ds...>;
+template <ptrdiff_t... Ds>
+using _exts = stdex::extents<Ds...>;
+
+using extents_test_types =
+  ::testing::Types<
+    std::tuple<_exts<10>, _sizes<>>,
+    std::tuple<_exts<stdex::dynamic_extent>, _sizes<10>>,
+    std::tuple<_exts<10, 3>, _sizes<>>,
+    std::tuple<_exts<stdex::dynamic_extent, 3>, _sizes<10>>,
+    std::tuple<_exts<10, stdex::dynamic_extent>, _sizes<3>>,
+    std::tuple<_exts<stdex::dynamic_extent, stdex::dynamic_extent>, _sizes<10, 3>>
+  >;
+
+TYPED_TEST_SUITE(TestExtents, extents_test_types);
+
+TYPED_TEST(TestExtents, rank) {
+  EXPECT_EQ(this->exts.rank(), this->static_sizes.size());
+}
+
+TYPED_TEST(TestExtents, dynamic_rank) {
+  EXPECT_EQ(this->exts.rank_dynamic(), this->dyn_sizes.size());
+}
+
+TYPED_TEST(TestExtents, static_extent) {
+  for (size_t r = 0; r < this->exts.rank(); ++r) {
+    EXPECT_EQ(this->exts.static_extent(r), this->static_sizes[r]);
+  }
+}
+
+TYPED_TEST(TestExtents, extent_method) {
+  size_t dyn_count = 0;
+  for (size_t r = 0; r < this->exts.rank(); ++r) {
+    bool is_dynamic = (this->exts.static_extent(r) == stdex::dynamic_extent);
+    EXPECT_EQ(this->exts.extent(r), is_dynamic ? this->dyn_sizes[dyn_count++] : this->static_sizes[r]);
+  }
+}
+
+TYPED_TEST(TestExtents, default_ctor) {
+  auto e = typename TestFixture::extents_type();
+  auto e2 = typename TestFixture::extents_type{};
+  EXPECT_EQ(e, e2);
+  for (size_t r = 0; r < e.rank(); ++r) {
+    bool is_dynamic = (e.static_extent(r) == stdex::dynamic_extent);
+    EXPECT_EQ(e.extent(r), is_dynamic ? 0 : e.static_extent(r));
+  }
+}
+
+TYPED_TEST(TestExtents, array_ctor) {
+  auto e = typename TestFixture::extents_type(this->dyn_sizes);
+  EXPECT_EQ(e, this->exts);
+}
+
+TYPED_TEST(TestExtents, copy_ctor) {
+  typename TestFixture::extents_type e { this->exts };
+  EXPECT_EQ(e, this->exts);
+}
+
+TYPED_TEST(TestExtents, copy_assign) {
+  typename TestFixture::extents_type e = this->exts;
```

Comment:
This is not testing the copy assignment operator.  It uses the copy constructor.

Reply:
Will fix in a follow up PR

Reply:
Thanks for this correction - just saw it.